File: problems/0289_game-of-life/solution_state_bits.java
class Solution {

    public int[][] gameOfLife(int[][] board) {
        int m = board.length;
        int n = board[0].length;
        int[][] dirs = { { -1, -1 }, { -1, 0 }, { -1, 1 }, { 0, -1 }, { 0, 1 }, { 1, -1 }, { 1, 0 }, { 1, 1 } };
        for (int r = 0; r < m; r++) {
            for (int c = 0; c < n; c++) {
                int live = 0;
                for (int[] d : dirs) {
                    int nr = r + d[0];
                    int nc = c + d[1];
                    if (nr >= 0 && nr < m && nc >= 0 && nc < n && (board[nr][nc] == 1 || board[nr][nc] == 2)) {
                        live++;
                    }
                }
                if (board[r][c] == 1 && (live < 2 || live > 3)) {
                    board[r][c] = 2; // live -> dead
                } else if (board[r][c] == 0 && live == 3) {
                    board[r][c] = 3; // dead -> live
                }
            }
        }
        for (int r = 0; r < m; r++) {
            for (int c = 0; c < n; c++) {
                board[r][c] = board[r][c] == 1 || board[r][c] == 3 ? 1 : 0;
            }
        }
        return board;
    }
}

File: problems/0201-0300/0289_life-one-generation/solution_state_bits.ts
function nextGeneration(board: number[][]): number[][] {
    const m = board.length;
    const n = board[0].length;
    const dirs: Array<[number, number]> = [
        [-1, -1],
        [-1, 0],
        [-1, 1],
        [0, -1],
        [0, 1],
        [1, -1],
        [1, 0],
        [1, 1],
    ];
    for (let r = 0; r < m; r++) {
        for (let c = 0; c < n; c++) {
            let live = 0;
            for (const [dr, dc] of dirs) {
                const nr = r + dr;
                const nc = c + dc;
                if (nr >= 0 && nr < m && nc >= 0 && nc < n && (board[nr][nc] === 1 || board[nr][nc] === 2)) {
                    live++;
                }
            }
            if (board[r][c] === 1 && (live < 2 || live > 3)) {
                board[r][c] = 2; // live -> dead
            } else if (board[r][c] === 0 && live === 3) {
                board[r][c] = 3; // dead -> live
            }
        }
    }
    for (let r = 0; r < m; r++) {
        for (let c = 0; c < n; c++) {
            board[r][c] = board[r][c] === 1 || board[r][c] === 3 ? 1 : 0;
        }
    }
    return board;
}
